### PR TITLE
Reduce concurrent macOS runners on PRs from 6 to 2 (and remove push-to-master triggers)

### LINKED ISSUE #4400: Limit of 5 concurrent macOS runners is causing development delays due to queued runners
In https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4353, we expanded our macOS testing to include both 12 and 14 in our test suite. We did this because `macos-12` is an Intel runner, while `macos-14` is an M1 runner, and thus theoretically running both would provide better coverage.

However, what I didn't realize at the time is that GitHub limits concurrent macOS runners (of any kind) to [5 on free plans](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits). (For some reason, I was under the impression that macOS 12 and macOS 14 would be treated as separate, rather than drawing from the same pool of concurrent runners).

Combined with https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4374, this really creates a bottleneck, since we fill up our macOS quota real quick. Merging a pull request, then immediately syncing another open PR with master will automatically fill up the queue and then some. Note the 3 separate "Queued" runs after 10 minutes:

image

I've found myself manually cancelling runs that I suspect won't reveal anything new, just to free up macOS runners for the jobs I _do_ care about.

It doesn't look like we can pay our way out of this limitation, either -- "Pro" and "Team" plans are limited to 5 concurrent macOS runners as well. So, we need to make sure we're careful about how we're allocating our 5 macOS runners at any given time.

----

On PRs, pushing a new commit will trigger 6 macOS runners (2 each for both Tests, batch_processing.sh, and Create Release), already exceeding the quota.

So, some possible suggestions:

1. Don't test both macOS 12 and 14 on PRs. 
   - On reflection, given that SCT requires Rosetta, I'm not sure that regularly testing on both Intel and Apple Silicon Macs actually provides much benefit. SCT's environment is sandboxed, so, apart from some obscure Rosetta-specific bugs, the testing will largely be the same, I think? Meaning, we should be okay to save, say, the macOS 12 test for `master` runs.
   - Not to mention, what will happen when GitHub deprecates the `macos-12` runners, anyway? Will macOS runners be Apple Silicon only? In such a future, we won't need to run two sets of runners anyway.
3. Somehow smartly combine existing workflows such that some tests are executed sequentially on a single macOS runner. This might result in a longer time spent on a single runner, but would more effectively use our quota for concurrent runners.
   - Probably more trouble than it's worth? I imagine it's just as good to have fast concurrent runners that free up quickly?
5. Speed up duplicated CI runs and/or reduce duplication of effort between "Create Release" runs and "Tests" runs
   - e.g. For the "Create Release" runs, only test the installation + `sct_check_dependencies` (rather than running the full test suite) on PRs, since the "Create Release" runs are mainly intended to test weird package/installation quirks that stem from freezing our requirements?
   - e.g. For the "Create Release" runs, only run the test suite if the installed requirements actually differ from the expected requirements (for the aforementioned reasons). Maybe we can use some sort of caching here, since we're already doing the work of saving the `requirements-freeze.txt` artifact between runs?
6. Skip running tests on `master` immediately after merge, and leave `master` tests for cron jobs only, to minimize the bottleneck stemming from cascades of merge/syncs.
7. ~~Reschedule the cron jobs to run at a time that's unlikely to clash with normal development. (I think right now they run midday, but they could be changed to run at, say, 6am/7am EST so that we get feedback when we clock in.~~ They already run at 7AM ET.

### PR DESCRIPTION
**Pull request by joshuacwnewton**
## Description

This PR solves the queuing issues we've faced due to the limit of 5 concurrent macOS runners.

Here, I've added the lowest-hanging-fruit suggestions from the issue description of https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4400. I've tried to preserve the highest-impact CI workflows (tests.yml, test-batch-processing.yml) while relegating the lower-impact CI workflows (create-release.yml, macos-14) to cron-only. I've also removed the trigger for merges/pushes to master. 

(See extended commit messages for further justification of each change.)

With these changes, we can now:

- Update PRs twice concurrently without running into queuing issues. (2x2 < 5)
- Merge and update PRs without running into queuing issues.

In the short term, this should allow us to quickly work through our backlog of approved PRs. (This need is also why I've mainly focused on uncontroversial, easy-to-review changes.) Long term, however, we may want to consider a more sophisticated approach to testing the create-release.yml workflow.

## Linked issues

Fixes #4400.
Fixes https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4374.
Maybe even addresses https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4034, since now that push-to-master triggers are gone, trains of PR merges are less of a hassle?

### CODE REVIEW

File: .github/workflows/create-release.yml
Comment (joshuacwnewton) [inline .github/workflows/create-release.yml:14]: There is perhaps a more nuanced approach we can take that would allow us to run the create-release.yml workflow on PRs without using additional macOS runners. For example, we could choose to only run the workflow if `requirements.txt` changes. We could even compare the `requirements-freeze.txt` artifact to a cached version, and only run the test suite if this file has changed. But, given the care required to implement such a solution, and given the need to unblock PRs ASAP, I figure we could try this in a follow-up PR.

Comment (mguaypaq) [reply .github/workflows/create-release.yml:14]: I think it's fine if we only run the create-release workflow as a daily cron job (or on-demand, if it's in the days leading up to a release).

File: .github/workflows/test-batch-processing.yml
Comment (mguaypaq) [inline .github/workflows/test-batch-processing.yml:1]: Same as for create-release, I think it's fine to run test-batch-processing only on a daily schedule.

File: .github/workflows/tests.yml
Comment (mguaypaq) [inline .github/workflows/tests.yml:7]: I agree with the logic here: if we just ran the test suite on an up-to-date PR before merging into master, there's almost certainly no need to re-run it on master. The very unlikely exceptions will get caught by the daily run anyway.

However, should we update all the `if: ${{ ... || github.event_name == 'push') }}` clauses to remove the references to push events, since they no longer exist?

File: .github/workflows/tests.yml
Comment (mguaypaq) [inline .github/workflows/tests.yml:265]: Just to double-check my understanding: the commit description for this change says that it's **disabling** the macos-14 runner, but these lines actually **enable** the macos-14 runner, right? Specifically, it becomes enabled for pull_request events?

I think it's still true that merging this PR would reduce the number of macos runners for pull_request events from 3 to 2.

**Review by mguaypaq — APPROVED** [review]
Thanks for this simplification of the CI runs, it should solve the immediate bottleneck problems.

File: .github/workflows/tests.yml
Comment (joshuacwnewton) [reply .github/workflows/tests.yml:265]: Ah! Sorry for the confusion... 

Without this line, the macos-14 runner would run unconditionally. Whereas, with this line, the macos-14 runner only runs for the triggers listed. And, since PRs aren't listed, it gets disabled for PRs.

File: .github/workflows/tests.yml
Comment (joshuacwnewton) [reply .github/workflows/tests.yml:7]: Ah, of course! Good catch.

File: .github/workflows/test-batch-processing.yml
Comment (joshuacwnewton) [reply .github/workflows/test-batch-processing.yml:1]: The main reason I kept the batch_processing test on PRs is because in the past, it's helped us to know when PR changes would affect the results of our processing pipelines.

Comment (joshuacwnewton) [reply .github/workflows/test-batch-processing.yml:1]: That said, batch_processing.sh still isn't the best at detecting cross-platform numerical discrepancies, due to https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3194.

File: .github/workflows/tests.yml
Comment (joshuacwnewton) [reply .github/workflows/tests.yml]: It's almost like a double-negative? It might be a little more clear if it was instead worded like this:

```
if ${{ github.event_name != 'pull_request' }}
```

That way it's clearer that it runs on every trigger _but_ PRs.

Comment (mguaypaq) [reply .github/workflows/tests.yml]: Oh wow, I read the diff backwards, sorry! For some reason I was convinced that these lines got removed instead of added :facepalm:

**Pull request merged**
